libnavigation-core/src/main/java/com/mapbox/navigation/core/routealternatives/OnlineRouteAlternativesSwitch.kt

```diff
@@ -0,0 +1,328 @@
+                    logE(LOG_CATEGORY) {
+                        "Not retryable error has occurred, " +
+                            "won't retry until new offline route is set"
+                    }
+                    return@coroutineScope null
+                }
+                is DirectionsRequestResult.ErrorResponse.RetryableErrorWithDelay -> {
+                    logI(LOG_CATEGORY) {
+                        "Error requesting route. " +
+                            "Applying additional delay of ${routeRequestResult.delayMilliseconds}"
+                    }
+                    delay(routeRequestResult.delayMilliseconds)
+                    retryLimiter.join()
+                }
+                else -> {
+                    logI(LOG_CATEGORY) { "failed to receive an online route, will retry later" }
+                    retryLimiter.join()
+                }
+            }
+        } finally {
+            retryLimiter.cancel()
+        }
+    }
+    null
+}
+
+private fun updateRouteOptionsOrNull(
+    primaryRoute: NavigationRoute,
+    routeProgress: RouteProgress,
+    locations: LocationMatcherResult
+): RouteOptions? {
+    val routeOptionsUpdateResult = RouteOptionsUpdater().update(
+        primaryRoute.routeOptions,
+        routeProgress,
+        locations
+    )
+    val newRouteOptions = when (routeOptionsUpdateResult) {
+        is RouteOptionsUpdater.RouteOptionsResult.Error -> {
+            logE(LOG_CATEGORY) { "Can't update route options: ${routeOptionsUpdateResult.error}" }
+            null
+        }
+        is RouteOptionsUpdater.RouteOptionsResult.Success -> {
+            routeOptionsUpdateResult.routeOptions
+        }
+    }
+    return newRouteOptions
+}
+
+private suspend fun requestRoutes(
+    routeOptions: RouteOptions,
+    accessToken: String,
+    urlSkuTokenProvider: UrlSkuTokenProvider,
+    connectTimeoutMilliseconds: Int,
+    readTimeoutMilliseconds: Int
+): DirectionsRequestResult {
+    val mapboxDirections = MapboxDirections.builder()
+        .routeOptions(routeOptions)
+        .accessToken(accessToken)
+        .interceptor {
+            val httpUrl = it.request().url
+            val skuUrl = urlSkuTokenProvider.obtainUrlWithSkuToken(httpUrl.toUrl())
+            val request = it.request().newBuilder().url(skuUrl).build()
+            it.withConnectTimeout(connectTimeoutMilliseconds, TimeUnit.MILLISECONDS)
+                .withReadTimeout(readTimeoutMilliseconds, TimeUnit.MILLISECONDS)
+                .proceed(request)
+        }
+        .build()
+    return suspendCancellableCoroutine { continuation ->
+        mapboxDirections.enqueueCall(object : Callback<DirectionsResponse> {
+            override fun onResponse(
+                call: Call<DirectionsResponse>,
+                response: Response<DirectionsResponse>
+            ) {
+                val body = response.body()
+                val result = if (body != null) {
```

Comment:
Does this mean it's successful? I think they can include body even when code is not 200. There are error messages.

Reply:
fair enough, will update tests